File: packages/native-helpers/linux-helper-py/linux_helper/evdev/monitor.py
"""Keyboard event monitoring via Linux evdev."""

import os
import struct
import sys
import threading
from datetime import datetime, timezone
from pathlib import Path

from ..stdout_writer import write_event
from .keycodes import is_modifier, get_modifier_flags, KEY_FN

# Linux input event constants
EV_KEY = 1
# struct input_event on 64-bit:
# uint64 sec, uint64 usec, uint16 type, uint16 code, int32 value
INPUT_EVENT_FORMAT = "QQHHi"
INPUT_EVENT_SIZE = struct.calcsize(INPUT_EVENT_FORMAT)

# Track pressed keys globally
pressed_keys: set[int] = set()
_lock = threading.Lock()


def get_pressed_keys() -> set[int]:
    with _lock:
        return pressed_keys.copy()


def _scan_keyboard_devices() -> list[str]:
    """Find keyboard devices by checking their key capabilities."""
    devices = []
    input_dir = Path("/dev/input")

    try:
        for entry in sorted(input_dir.iterdir()):
            if not entry.name.startswith("event"):
                continue

            caps_path = Path(f"/sys/class/input/{entry.name}/device/capabilities/key")
            try:
                caps = caps_path.read_text().strip()
                if not caps or caps == "0":
                    continue

                # Count set bits - real keyboards have many key capabilities
                total_bits = sum(
                    bin(int(part, 16)).count("1") for part in caps.split()
                )
                if total_bits > 20:
                    devices.append(str(entry))
            except (OSError, ValueError):
                continue
    except OSError as e:
        sys.stderr.write(f"Failed to scan input devices: {e}\n")

    return devices


def _emit_key_event(event_type: str, key_code: int) -> None:
    flags = get_modifier_flags(pressed_keys)
    event = {
        "type": event_type,
        "payload": {
            "keyCode": key_code,
            "key": None,
            "code": None,
            "altKey": flags["altKey"],
            "ctrlKey": flags["ctrlKey"],
            "shiftKey": flags["shiftKey"],
            "metaKey": flags["metaKey"],
            "fnKeyPressed": KEY_FN in pressed_keys,
        },
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
    write_event(event)


def _monitor_device(device_path: str) -> None:
    """Monitor a single evdev device for key events."""
    while True:
        try:
            fd = os.open(device_path, os.O_RDONLY)
        except OSError as e:
            sys.stderr.write(f"Cannot open {device_path}: {e}\n")
            return

        try:
            while True:
                data = os.read(fd, INPUT_EVENT_SIZE * 64)
                if not data:
                    break

                offset = 0
                while offset + INPUT_EVENT_SIZE <= len(data):
                    _sec, _usec, ev_type, code, value = struct.unpack_from(
                        INPUT_EVENT_FORMAT, data, offset
                    )
                    offset += INPUT_EVENT_SIZE

                    if ev_type != EV_KEY:
                        continue

                    with _lock:
                        if value == 1:  # Key down
                            pressed_keys.add(code)
                            if is_modifier(code):
                                _emit_key_event("flagsChanged", code)
                            else:
                                _emit_key_event("keyDown", code)
                        elif value == 0:  # Key up
                            pressed_keys.discard(code)
                            if is_modifier(code):
                                _emit_key_event("flagsChanged", code)
                            else:
                                _emit_key_event("keyUp", code)
                        elif value == 2:  # Key repeat
                            _emit_key_event("keyDown", code)
        except OSError as e:
            sys.stderr.write(f"Error reading {device_path}: {e}\n")
        finally:
            os.close(fd)

        import time
        sys.stderr.write(f"Device {device_path} closed, retrying in 5s\n")
        time.sleep(5)


def start_keyboard_monitor() -> None:
    """Start monitoring all keyboard devices in background threads."""
    devices = _scan_keyboard_devices()
    if not devices:
        sys.stderr.write(
            "No keyboard devices found. Ensure user is in 'input' group.\n"
        )
        return

    sys.stderr.write(f"Monitoring keyboard devices: {', '.join(devices)}\n")
    for device in devices:
        t = threading.Thread(target=_monitor_device, args=(device,), daemon=True)
        t.start()

File: packages/native-helpers/linux-helper-ts/src/evdev/monitor.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { writeEvent } from "../rpc/stdout-writer.js";
import { getModifierFlags, isModifierKeyCode } from "./keycodes.js";

// Linux input event constants
const EV_KEY = 1;
// struct input_event on 64-bit: { uint64 sec, uint64 usec, uint16 type, uint16 code, int32 value }
const INPUT_EVENT_SIZE = 24;

// Track pressed keys globally for modifier state
const pressedKeys = new Set<number>();

export function getPressedKeys(): Set<number> {
  return pressedKeys;
}

function scanKeyboardDevices(): string[] {
  const devices: string[] = [];
  try {
    const inputDir = "/dev/input";
    const entries = fs.readdirSync(inputDir);
    for (const entry of entries) {
      if (!entry.startsWith("event")) continue;
      const devicePath = path.join(inputDir, entry);

      // Check read access first
      try {
        fs.accessSync(devicePath, fs.constants.R_OK);
      } catch {
        continue; // Skip devices we can't read
      }

      // Check if this is a keyboard by reading capabilities
      const capsPath = `/sys/class/input/${entry}/device/capabilities/key`;
      try {
        const caps = fs.readFileSync(capsPath, "utf-8").trim();
        if (caps && caps !== "0") {
          const parts = caps.split(" ");
          const totalBits = parts.reduce((sum: number, hex: string) => {
            // Use BigInt to handle 64-bit hex values correctly
            let count = 0;
            let n = BigInt(`0x${hex}`);
            while (n) {
              count += Number(n & 1n);
              n >>= 1n;
            }
            return sum + count;
          }, 0);
          // Real keyboards have many key capabilities (>20 bits set)
          if (totalBits > 20) {
            devices.push(devicePath);
          }
        }
      } catch {
        // Skip devices we can't inspect
      }
    }
  } catch (err) {
    process.stderr.write(`Failed to scan input devices: ${err}\n`);
  }
  return devices;
}

function parseInputEvent(
  buf: Buffer,
  offset: number,
): { type: number; code: number; value: number } | null {
  if (buf.length - offset < INPUT_EVENT_SIZE) return null;
  // Skip timestamp (16 bytes), read type, code, value
  const type = buf.readUInt16LE(offset + 16);
  const code = buf.readUInt16LE(offset + 18);
  const value = buf.readInt32LE(offset + 20);
  return { type, code, value };
}

function emitKeyEvent(
  eventType: "keyDown" | "keyUp" | "flagsChanged",
  keyCode: number,
): void {
  const flags = getModifierFlags(pressedKeys);
  const event = {
    type: eventType,
    payload: {
      keyCode,
      altKey: flags.altKey,
      ctrlKey: flags.ctrlKey,
      shiftKey: flags.shiftKey,
      metaKey: flags.metaKey,
      fnKeyPressed: pressedKeys.has(464), // KEY_FN
    },
    timestamp: new Date().toISOString(),
  };
  writeEvent(event);
}

function monitorDevice(devicePath: string): void {
  let fd: number;
  try {
    fd = fs.openSync(devicePath, "r");
  } catch (err) {
    process.stderr.write(`Cannot open ${devicePath}: ${err}\n`);
    return;
  }

  const buf = Buffer.alloc(INPUT_EVENT_SIZE * 64);

  function readLoop(): void {
    fs.read(fd, buf, 0, buf.length, null, (err, bytesRead) => {
      if (err) {
        process.stderr.write(`Error reading ${devicePath}: ${err.message}\n`);
        try { fs.closeSync(fd); } catch { /* ignore */ }
        // Retry after delay
        setTimeout(() => monitorDevice(devicePath), 5000);
        return;
      }

      if (bytesRead === 0) {
        try { fs.closeSync(fd); } catch { /* ignore */ }
        process.stderr.write(`Device ${devicePath} EOF, retrying in 5s\n`);
        setTimeout(() => monitorDevice(devicePath), 5000);
        return;
      }

      let offset = 0;
      while (offset + INPUT_EVENT_SIZE <= bytesRead) {
        const event = parseInputEvent(buf, offset);
        offset += INPUT_EVENT_SIZE;
        if (!event || event.type !== EV_KEY) continue;

        const { code, value } = event;

        if (value === 1) {
          pressedKeys.add(code);
          emitKeyEvent(isModifierKeyCode(code) ? "flagsChanged" : "keyDown", code);
        } else if (value === 0) {
          pressedKeys.delete(code);
          emitKeyEvent(isModifierKeyCode(code) ? "flagsChanged" : "keyUp", code);
        } else if (value === 2) {
          emitKeyEvent("keyDown", code);
        }
      }

      // Continue reading
      readLoop();
    });
  }

  readLoop();
}

export function startKeyboardMonitor(): void {
  const devices = scanKeyboardDevices();
  if (devices.length === 0) {
    const groups = fs.readFileSync("/proc/self/status", "utf-8").match(/^Groups:\s*(.*)$/m)?.[1] ?? "";
    const inputGid = fs.readFileSync("/etc/group", "utf-8").match(/^input:x:(\d+):/m)?.[1];
    const hint = inputGid && !groups.split(/\s+/).includes(inputGid) ? " User is NOT in 'input' group. Run: sudo usermod -aG input $USER (then re-login)" : "";
    process.stderr.write(`No readable keyboard devices found.${hint}\n`);
    return;
  }

  process.stderr.write(`Monitoring keyboard devices: ${devices.join(", ")}\n`);
  for (const device of devices) {
    monitorDevice(device);
  }
}
